media-proxy/include/mesh/st2110rx.h

```diff
@@ -0,0 +1,155 @@
+#ifndef ST2110RX_H
+#define ST2110RX_H
+
+#include "st2110.h"
+
+namespace mesh::connection {
+
+/**
+ * ST2110Rx
+ *
+ * Base abstract class of ST2110Rx. ST2110_20Rx/ST2110_22Rx/ST2110_30Rx
+ * inherit this class.
+ */
+template <typename FRAME, typename HANDLE, typename OPS> class ST2110Rx : public ST2110 {
+  public:
+    ST2110Rx()
+    {
+        _kind = Kind::receiver;
+
+        _handle = nullptr;
+        _ops = {0};
+        _transfer_size = 0;
+    }
+    ~ST2110Rx() { shutdown(_ctx); }
+
+  protected:
+    HANDLE _handle;
+    OPS _ops;
+    size_t _transfer_size;
+    std::jthread _frame_thread_handle;
+    context::Context _ctx;
+
+    virtual FRAME *get_frame(HANDLE) = 0;
+    virtual int put_frame(HANDLE, FRAME *) = 0;
+    virtual HANDLE create_session(mtl_handle, OPS *) = 0;
+    virtual int close_session(HANDLE) = 0;
+
+    Result on_establish(context::Context &ctx) override
+    {
+        _ctx = context::WithCancel(ctx);
+        _stop = false;
+
+        _handle = create_session(_st, &_ops);
+        if (!_handle) {
+            log::error("Failed to create session");
+            set_state(ctx, State::closed);
+            return set_result(Result::error_general_failure);
+        }
+
+        /* Start MTL session thread. */
+        try {
+            _frame_thread_handle = std::jthread(&ST2110Rx::frame_thread, this);
+        } catch (const std::system_error &e) {
+            log::error("Failed to create thread");
+            set_state(ctx, State::closed);
+            return set_result(Result::error_out_of_memory);
+        }
+
+        set_state(ctx, State::active);
+        return set_result(Result::success);
+    }
+
+    Result on_shutdown(context::Context &ctx) override
+    {
+        _ctx.cancel();
+
+        _frame_thread_handle.join();
+
+        if (_handle) {
+            close_session(_handle);
+            _handle = nullptr;
+        }
+        set_state(ctx, State::closed);
+        return set_result(Result::success);
+    };
+
+    virtual void on_delete(context::Context &ctx) override {}
+
+  private:
+    void frame_thread()
```

Comment:
This needs to be reworked to avoid using the mutex and the condition variable. Use atomics only.